File: algoritms/recommendedgoods_cos.py
import math

mentions = dict()

mentions['alex'] = dict()
mentions['alex'][1] = 5.0
mentions['alex'][2] = 3.0
mentions['alex'][5] = 4.0

mentions['rob'] = dict()
mentions['rob'][1] = 4.0
mentions['rob'][6] = 1.0
mentions['rob'][8] = 2.0
mentions['rob'][9] = 3.0

mentions['bob'] = dict()
mentions['bob'][2] = 5.0
mentions['bob'][3] = 5.0

mentions['david'] = dict()
mentions['david'][3] = 4.0
mentions['david'][4] = 3.0
mentions['david'][6] = 2.0
mentions['david'][7] = 1.0

print(mentions)

def distCosine(vecA, vecB):
    print(vecA, vecB)
    def dotProduct(vecA, vecB):
        d = 0.0
        for dim in vecA:
            if dim in vecB:
                #print(dim, vecB)
                d += vecA[dim]*vecB[dim]
        return d
    return dotProduct(vecA,vecB) / math.sqrt(dotProduct(vecA,vecA)) / math.sqrt(dotProduct(vecB,vecB))

def makeRecommendation(userID, userRates, nBestUsers, nBestProducts):
    matches = [(u, distCosine(userRates[userID], userRates[u])) for u in userRates if u != userID]
    print(matches)
    bestMatches = sorted(matches, key=lambda x: (x[1], x[0]), reverse=True)[:nBestUsers]
    print("Most suitable for '%s' users:" % userID)
    for line in bestMatches:
        print("  UserID: %6s  coeff: %6.4f" % (line[0], line[1])) 
          
    sim = dict()
    sim_all = sum([x[1] for x in bestMatches])
    print(sim_all)
    bestMatches = dict([x for x in bestMatches if x[1] > 0.0])
           
    for relatedUser in bestMatches:
        for product in userRates[relatedUser]:
            if not product in userRates[userID]:
                if not product in sim:
                    sim[product] = 0.0
                sim[product] += userRates[relatedUser][product] * bestMatches[relatedUser]
                
    for product in sim:
        sim[product] /= sim_all
        
    bestProducts = sorted(sim.items(), key=lambda x: (x[1], x[0]), reverse=True)[:nBestProducts]
    print("Most suitable products:")
    for prodInfo in bestProducts:
        print("  ProductID: %6s  suitable coeff: %6.4f" % (prodInfo[0], prodInfo[1]))   
    return [(x[0], x[1]) for x in bestProducts]

rec = makeRecommendation('rob', mentions, 5, 5)

File: algoritms/recommendedgoods_cos.php
<?php 

$mentions = array();

$mentions['alex'] = array();
$mentions['alex'][1] = 5.0;
$mentions['alex'][2] = 3.0;
$mentions['alex'][5] = 4.0;

$mentions['rob'] = array();
$mentions['rob'][1] = 4.0;
$mentions['rob'][6] = 1.0;
$mentions['rob'][8] = 2.0;
$mentions['rob'][9] = 3.0;

$mentions['bob'] = array();
$mentions['bob'][2] = 5.0;
$mentions['bob'][3] = 5.0;

$mentions['david'] = array();
$mentions['david'][3] = 4.0;
$mentions['david'][4] = 3.0;
$mentions['david'][6] = 2.0;
$mentions['david'][7] = 1.0;

//print_r($mentions);
//echo "\n";

function distCosine($vecA, $vecB) {
    if (!function_exists('dotProduct')) {
        function dotProduct($vecA, $vecB) {
            $d = 0.0;
            foreach ($vecA as $dim => $value) {
                //echo "$dim $value \n";
                if(isset($vecB[$dim])) {
                    $d += $vecA[$dim] * $vecB[$dim];
                }
            }
            return $d;
        }
    }
    return dotProduct($vecA, $vecB) / sqrt(dotProduct($vecA, $vecA)) / sqrt(dotProduct($vecB, $vecB));
}

function makeRecommendation($userID, $userRates, $nBestUsers, $nBestProducts) {
    
    $matches = array();
    
    foreach ($userRates as $u => $value) {
        if($u != $userID){
            $matches[$u] = distCosine($userRates[$userID], $userRates[$u]);
        }
    }
    
    function cmp($a, $b) { if ($a == $b) { return 0; } return ($a > $b) ? -1 : 1; }
    uasort($matches, 'cmp');
    
    echo "Most suitable for $userID users: \n";
    $i=0;
    foreach ($matches as $key => $value) {
        echo "UserID: $key coeff: $value \n";
        if (++$i==$nBestUsers)
            break;
    }
    
    $sim = array();
    $sim_all = array_sum($matches);
    foreach ($matches as $relatedUser => $value) {
        if($value == 0) { continue; }
        
        foreach ($userRates[$relatedUser] as $product => $value) {
            if(!isset($sim[$product])) {
                $sim[$product] = 0.0;
            }
            $sim[$product] += $userRates[$relatedUser][$product] * $matches[$relatedUser];
        } 
    }
    
    foreach ($sim as $product => $value) {
        $sim[$product] /= $sim_all;
    }
    
    echo "Most suitable products:";
    uasort($sim, 'cmp');
    $i=0;
    foreach ($sim as $key => $value) {
        echo "ProductID: $key suitable coeff: $value \n";
        if (++$i==$nBestProducts) 
            break;
    }
    
    //print_r($matches);
    //echo "\n";
}

makeRecommendation('rob', $mentions, 5, 5);

?>